### PR TITLE
[Fixes #255] Change object type yielded to ActiveSupport::Subscribers

### LINKED ISSUE #255: Instrumentation yields unexpected object to "rack.attack" subscriptions.
Based on the [ActiveSupport documentation](http://api.rubyonrails.org/classes/ActiveSupport/Notifications.html#module-ActiveSupport::Notifications-label-Subscribers):

```
ActiveSupport::Notifications.subscribe('render') do |name, start, finish, id, payload|
  name    # => String, name of the event (such as 'render' from above)
  start   # => Time, when the instrumented block started execution
  finish  # => Time, when the instrumented block ended execution
  id      # => String, unique ID for this notification
  payload # => Hash, the payload
end
```

The block parameter `payload` is required to be a `Hash`.

However, Rack::Attack [defines `instrument` as](https://github.com/Domenoth/rack-attack/blob/master/lib/rack/attack.rb#L99):

```
    def instrument(req)
      notifier.instrument('rack.attack', req) if notifier
    end
```

And it passes a `Rack::Attack::Request` object instead of a `Hash`.

This causes issues when using [`attach_to`](http://api.rubyonrails.org/classes/ActiveSupport/Subscriber.html#method-c-attach_to):

```
  1) Error:
Rack::Attack.instrument::with throttling#test_0001_should instrument without error:
NoMethodError: undefined method `merge!' for #<Rack::Attack::Request:0x00007fdc329b1598>
```

The `Hash` method `merge!` is called on `payload` as part of the [`ActiveSupport::Subscriber` internal method `finish`](https://github.com/rails/rails/blob/a43e7706be907e62fad4ffe5f80ab9532c51dbca/activesupport/lib/active_support/subscriber.rb#L96).

### PR DESCRIPTION
**Pull request by Domenoth**
https://github.com/kickstarter/rack-attack/issues/255

Change the object type from instances of type Rack::Attack::Request to
instances of type Hash. (`req` becomes `:request => req`).

Fixes: #255

### CODE REVIEW

File: lib/rack/attack.rb [hunk 1]
```diff
@@ -97,7 +97,7 @@ def tracked?(req)
     end
 
     def instrument(req)
-      notifier.instrument('rack.attack', req) if notifier
+      notifier.instrument('rack.attack', :request => req) if notifier
```
Comment (lmansur) [inline lib/rack/attack.rb]: We should use Ruby 1.9 hash syntax moving forward.
This would become `request: req`

**Comment by Domenoth** [comment]
Switched to Ruby 1.9 hash syntax and fixed some tests that broke when applied to the newest master. 👍 

Let me know if there's more I can do. 😀

**Comment by lmansur** [comment]
Looks good to me :+1:

**Comment by grzuy** [comment]
@Domenoth Thank you very much for the pull request :-) :+1:

**Comment by grzuy** [comment]
Thank you @lmansur for the code review also

File: spec/rack_attack_instrumentation_spec.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    before do
+      @period = 60 # Use a long period; failures due to cache key rotation less likely
+      Rack::Attack.cache.store = ActiveSupport::Cache::MemoryStore.new
+      Rack::Attack.throttle('ip/sec', limit: 1, period: @period) { |req| req.ip }
+    end
+
+    describe "with throttling" do
+      before do
+        ActiveSupport::Notifications.stub(:notifier, ActiveSupport::Notifications::Fanout.new) do
+          CustomSubscriber.attach_to("attack")
```
[the remark this answers is not in the document]
Comment (grzuy) [reply spec/rack_attack_instrumentation_spec.rb]: FTR I created issue https://github.com/kickstarter/rack-attack/issues/371

**Pull request merged**